src/main/java/bio/ferlab/fhir/Fhavro.java

```diff
@@ -0,0 +1,265 @@
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.commons.io.IOUtils;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.BaseResource;
+import org.hl7.fhir.r4.model.DomainResource;
+import org.hl7.fhir.r4.model.StructureDefinition;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Fhavro is an open-source library for the standardization and conversion of FHIR v4.0.1 data to the Apache Avro format.
+ *
+ * @author https://github.com/Ferlab-Ste-Justine
+ */
+public class Fhavro {
+
+    private static final FhirContext fhirContext;
+
+    Fhavro() {
+    }
+
+    static {
+        fhirContext = FhirContext.forR4();
+    }
+
+    /**
+     * Convert a FHIR v4.0.1 Resource in a Generic Record based on a predefined Schema.
+     *
+     * @param resource The FHIR resource to be converted.
+     * @param schema   The Schema that represents the FHIR resource
+     * @return The GenericRecord for this FHIR resource
+     */
+    @NotNull
+    public static GenericData.Record convertResourceToGenericRecord(DomainResource resource, Schema schema) {
+        return FhirAvroConverter.readResource(resource, schema);
+    }
+
+    /**
+     * Convert a Generic Record representing a FHIR v4.0.1 Resource based on a predefined schema.
+     *
+     * @param genericRecord The GenericRecord representing a FHIR resource
+     * @param schema        The Schema that represents the FHIR resource
+     * @param name          The name of the FHIR resource (e.g: Patient, Account, etc).
+     * @return The FHIR resource
+     */
+    @NotNull
+    public static <T extends DomainResource> T convertGenericRecordToResource(GenericRecord genericRecord, Schema schema, String name) {
+        return AvroFhirConverter.readGenericRecord(genericRecord, schema, name);
+    }
+
+    /**
+     * Load the Schema by its name. The mode will determine where the schema will be fetched from.
+     * If the mode is DEFAULT or SIMPLE, the schema will be searched within the Resources root folder of this library.
+     * Else, if the mode is ADVANCED, the schema is searched from the Root of the calling project by its name.
+     * <p>
+     * SchemaMode:
+     * <p>SIMPLE/DEFAULT: ./src/resources/schemas/'schemaMode'/'schemaName'
+     * <p>ADVANCED: ./'schemaName'
+     *
+     * @param schemaName The name of the schema.
+     * @param schemaMode The Schema Mode dictating where to search
+     * @return The Schema
+     */
+    @NotNull
+    public static Schema loadSchema(String schemaName, SchemaMode schemaMode) {
+        schemaName = ConverterUtils.formatSchemaName(schemaName);
+
+        try {
+            InputStream inputStream;
```

Comment:
`InputStream` must be closed explicitly or with a try-with-resources

Reply:
https://www.baeldung.com/java-stream-close

Reply:
Good point, this makes me question my sanity.